pkg/cmd/pulumi/state_repair.go

```diff
@@ -0,0 +1,289 @@
+	// can be refactored.
+	backend.DisableIntegrityChecking = true
+	defer func() { backend.DisableIntegrityChecking = false }()
+
+	displayOpts := display.Options{
+		Color: cmd.Args.Colorizer,
+	}
+	s, err := requireStack(
+		ctx,
+		cmd.Workspace,
+		cmd.LoginManager,
+		cmd.Args.Stack,
+		stackOfferNew,
+		displayOpts,
+	)
+	if err != nil {
+		return err
+	}
+
+	snap, err := s.Snapshot(ctx, stack.DefaultSecretsProvider)
+	if err != nil {
+		return err
+	} else if snap == nil {
+		return nil
+	}
+
+	sink := diag.DefaultSink(cmd.Stdout, cmd.Stderr, diag.FormatOptions{
+		Color: cmd.Args.Colorizer,
+	})
+
+	// If the snapshot is already valid, we won't touch it.
+	initialErr := snap.VerifyIntegrity()
+	if initialErr == nil {
+		sink.Infof(diag.RawMessage("" /*urn*/, "The snapshot is already valid; skipping repair"))
+		return nil
+	}
+
+	// The snapshot isn't valid, so we should proceed. If the user has passed the --yes flag, we'll skip the confirmation
+	// prompt and proceed directly to the repair.
+	if !cmd.Args.Yes {
+		yes := "yes"
+		no := "no"
+		msg := "This command will edit your stack's state directly. Confirm?"
+		options := []string{yes, no}
+
+		switch response := promptUser(
+			msg, options, no, cmd.Args.Colorizer,
+			survey.WithStdio(cmd.Stdin, cmd.Stdout, cmd.Stderr),
+		); response {
+		case yes:
+			// Continue.
+		case no:
+			sink.Infof(diag.RawMessage("" /*urn*/, "Confirmation denied, not proceeding with state repair"))
+			return nil
+		}
+	}
+
+	// Sorting the snapshot could fail due to cycles or e.g. unparseable provider references. In those cases, manual
+	// repair is likely the only option, so we'll print a help banner to guide the user through that and invite them
+	// to file a report so that we can learn about how they ended up with such a state.
+	err = snap.Toposort()
+	if err != nil {
+		sink.Errorf(diag.RawMessage("" /*urn*/, cmd.manualRepairError(initialErr, err)))
+
+		// We've already taken care of printing an error message, so we'll wrap the error we return in a Bail so that
+		// callers higher up the stack don't print anything of their own.
+		return result.BailError(err)
+	}
+
+	snap.Prune()
+
+	// In the case that we complete repairs (sorting, pruning and so on) but the snapshot is still invalid, we'll
+	// produce a banner that helps the user conduct a manual repair but also includes both errors, so that if they
+	// file a report we can hopefully diagnose both what caused the invalid snapshot but also why we failed to repair
+	// it (assuming a repair was possible).
+	err = snap.VerifyIntegrity()
+	if err != nil {
+		sink.Errorf(diag.RawMessage("" /*urn*/, cmd.manualRepairError(initialErr, err)))
+
+		// We've already taken care of printing an error message, so we'll wrap the error we return in a Bail so that
+		// callers higher up the stack don't print anything of their own.
+		return result.BailError(initialErr)
```

Comment:
Wrapping it in a bail error will also make the command return a 0 exit code, or am I misremembering that?  I don't think that's correct here in that case.

Reply:
Ooh good question. I don't _think_ so -- https://github.com/pulumi/pulumi/blob/master/pkg/cmd/pulumi/errors.go#L63; it seems that bails only affect diagnostic output. Manual testing seems to confirm this too.

Reply:
Ah thanks, I was misremembering how exactly this works :)  Thanks for checking!